## Import Libaries

In [1]:
import pandas as pd
import random

## Load dataset

In [2]:
!ls

README.md
dataset
preprocess_data.ipynb
report


In [3]:
!ls dataset

marketing_data.csv
marketing_data_final.csv


In [4]:
df = pd.read_csv('dataset/marketing_data.csv')
df.head(5)

ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
0   1826        1970  Graduation       Divorced  $84,835.00         0   
1      1        1961  Graduation         Single  $57,091.00         0   
2  10476        1958  Graduation        Married  $67,267.00         0   
3   1386        1967  Graduation       Together  $32,474.00         1   
4   5371        1989  Graduation         Single  $21,474.00         1   

   Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
0         0   6/16/2014        0       189  ...                  6   
1         0   6/15/2014        0       464  ...                  7   
2         1   5/13/2014        0       134  ...                  5   
3         1   5/11/2014        0        10  ...                  2   
4         0    4/8/2014        0         6  ...                  2   

   NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  AcceptedCmp1  \
0                  1             0             0             0             0   
1                  5             0             0             0             0   
2                  2             0             0             0             0   
3                  7             0             0             0             0   
4                  7             1             0             0             0   

   AcceptedCmp2  Response  Complain  Country  
0             0         1         0       SP  
1             1         1         0       CA  
2             0         0         0       US  
3             0         0         0      AUS  
4             0         1         0       SP  

[5 rows x 28 columns]

In [5]:
df.sort_values(by='ID', inplace=True)
df.head(5)

ID  Year_Birth   Education Marital_Status      Income   Kidhome  \
1503   0        1985  Graduation        Married  $70,951.00         0   
1      1        1961  Graduation         Single  $57,091.00         0   
1956   9        1975      Master         Single  $46,098.00         1   
1311  13        1947         PhD          Widow  $25,358.00         0   
1834  17        1971         PhD        Married  $60,491.00         0   

      Teenhome Dt_Customer  Recency  MntWines  ...  NumStorePurchases  \
1503         0    5/4/2013       66       239  ...                  9   
1            0   6/15/2014        0       464  ...                  7   
1956         1   8/18/2012       86        57  ...                  2   
1311         1   7/22/2013       57        19  ...                  3   
1834         1    9/6/2013       81       637  ...                  7   

      NumWebVisitsMonth  AcceptedCmp3  AcceptedCmp4  AcceptedCmp5  \
1503                  1             0             0             0   
1                     5             0             0             0   
1956                  8             0             0             0   
1311                  6             0             0             0   
1834                  5             0             0             0   

      AcceptedCmp1  AcceptedCmp2  Response  Complain  Country  
1503             0             0         0         0       SA  
1                0             1         1         0       CA  
1956             0             0         0         0       CA  
1311             0             0         0         0       SA  
1834             0             0         0         0       SP  

[5 rows x 28 columns]

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2240 entries, 1503 to 261
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   ID                   2240 non-null   int64 
 1   Year_Birth           2240 non-null   int64 
 2   Education            2240 non-null   object
 3   Marital_Status       2240 non-null   object
 4    Income              2216 non-null   object
 5   Kidhome              2240 non-null   int64 
 6   Teenhome             2240 non-null   int64 
 7   Dt_Customer          2240 non-null   object
 8   Recency              2240 non-null   int64 
 9   MntWines             2240 non-null   int64 
 10  MntFruits            2240 non-null   int64 
 11  MntMeatProducts      2240 non-null   int64 
 12  MntFishProducts      2240 non-null   int64 
 13  MntSweetProducts     2240 non-null   int64 
 14  MntGoldProds         2240 non-null   int64 
 15  NumDealsPurchases    2240 non-null   int64 
 16  NumW

Ta thấy số lượng sample là 2240 dòng, riêng cột Income chỉ có 2216 dòng. Ta tiến hành kiểm tra các giá trị **NaN** rồi chọn phương hướng tiếp cận cho số lượng dòng của các cột như nhau

## Rename Column

In [7]:
df.rename(columns={'ID':'User_ID'}, inplace=True)
df.rename(columns={' Income ':'Income'}, inplace=True)
df.rename(columns={'Dt_Customer': 'Date_added'}, inplace = True)
df.rename(columns={'MntGoldProds': 'MntGoldProducts'}, inplace = True)

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2240 entries, 1503 to 261
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   User_ID              2240 non-null   int64 
 1   Year_Birth           2240 non-null   int64 
 2   Education            2240 non-null   object
 3   Marital_Status       2240 non-null   object
 4   Income               2216 non-null   object
 5   Kidhome              2240 non-null   int64 
 6   Teenhome             2240 non-null   int64 
 7   Date_added           2240 non-null   object
 8   Recency              2240 non-null   int64 
 9   MntWines             2240 non-null   int64 
 10  MntFruits            2240 non-null   int64 
 11  MntMeatProducts      2240 non-null   int64 
 12  MntFishProducts      2240 non-null   int64 
 13  MntSweetProducts     2240 non-null   int64 
 14  MntGoldProducts      2240 non-null   int64 
 15  NumDealsPurchases    2240 non-null   int64 
 16  NumW

## Categorical value column Country

In [9]:
df['Country'].unique()

array(['SA', 'CA', 'SP', 'GER', 'IND', 'US', 'AUS', 'ME'], dtype=object)

In [10]:
value_to_replace = {'SA': 'South Africa', 
                    'CA': 'Canada', 
                    'SP': 'Spain', 
                    'GER': 'Germany', 
                    'IND': 'India', 
                    'US': 'United States',
                    'AUS': 'Australia',
                   'ME': 'Montenegro'}
df['Country'].replace(value_to_replace, inplace=True)

In [11]:
df['Country'].head(5)

1503    South Africa
1             Canada
1956          Canada
1311    South Africa
1834           Spain
Name: Country, dtype: object

## Format datatime
Ta cần chuyển cột **Dt_Customer** về dạng Datatime theo format **yyyymmdd**

In [12]:
df['Date_added'] = pd.to_datetime(df['Date_added'])
df['Date_added'].head(5)

1503   2013-05-04
1      2014-06-15
1956   2012-08-18
1311   2013-07-22
1834   2013-09-06
Name: Date_added, dtype: datetime64[ns]

## Xử lý dữ liệu NaN
Tiến hành kiểm tra dữ liệu NaN cột Income và xử lý theo một trong những cách sau đây:
1. Drop tất cả những dòng mang dữ liệu NaN
2. Thế tất cả những dữ liệu NaN bằng mean dữ liệu

Nhóm mình quyết định sẽ xử lý theo cách thứ nhất để đảm bảo tính nguyên thủy cho dữ liệu

In [13]:
# Chuyển đổi kiểu dữ liệu cột income
df["Income"] = df["Income"].str.replace("$","").str.replace(',', '')
df["Income"] = df["Income"].astype(float)

C:\Users\NGHOAN~1\AppData\Local\Temp/ipykernel_3452/3242172660.py:2: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  df["Income"] = df["Income"].str.replace("$","").str.replace(',', '')


In [14]:
df['Income'].head(5)

1503    70951.0
1       57091.0
1956    46098.0
1311    25358.0
1834    60491.0
Name: Income, dtype: float64

In [15]:
# Kiểm tra số lượng NaN value của cột Income
df['Income'].isna().sum()

24

In [16]:
# Loại bỏ tất cả các dòng có chứa dữ liệu NaN
df.dropna(axis='index', how='any', inplace=True)

In [17]:
# Kiểm tra
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2216 entries, 1503 to 261
Data columns (total 28 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   User_ID              2216 non-null   int64         
 1   Year_Birth           2216 non-null   int64         
 2   Education            2216 non-null   object        
 3   Marital_Status       2216 non-null   object        
 4   Income               2216 non-null   float64       
 5   Kidhome              2216 non-null   int64         
 6   Teenhome             2216 non-null   int64         
 7   Date_added           2216 non-null   datetime64[ns]
 8   Recency              2216 non-null   int64         
 9   MntWines             2216 non-null   int64         
 10  MntFruits            2216 non-null   int64         
 11  MntMeatProducts      2216 non-null   int64         
 12  MntFishProducts      2216 non-null   int64         
 13  MntSweetProducts     2216 non-n

Ta thấy 24 dòng mang giá trị **NaN** đã được loại bỏ, dữ liệu còn lại tổng cộng 2216 dòng và 28 column

## Add new features
Sau khi trải qua các bước làm sạch dữ liệu, ta sẽ cần thêm vào những cột cần thiết để thiết kế data warehouse

In [18]:
# Thêm cột total_campaign_accepted: tổng số lượng marketing campaign đã chấp nhận
campaigns_cols = [col for col in df.columns if 'Cmp' in col] + ['Response']
df['Total_Campaigns_Accepted'] = df[campaigns_cols].sum(axis=1)

In [19]:
# Làm mịn dữ liệu về datetime
df['Year_added'] = pd.DatetimeIndex(df['Date_added']).year
df['Month_added'] = pd.DatetimeIndex(df['Date_added']).month
df['Day_added'] = pd.DatetimeIndex(df['Date_added']).day

In [20]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2216 entries, 1503 to 261
Data columns (total 32 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   User_ID                   2216 non-null   int64         
 1   Year_Birth                2216 non-null   int64         
 2   Education                 2216 non-null   object        
 3   Marital_Status            2216 non-null   object        
 4   Income                    2216 non-null   float64       
 5   Kidhome                   2216 non-null   int64         
 6   Teenhome                  2216 non-null   int64         
 7   Date_added                2216 non-null   datetime64[ns]
 8   Recency                   2216 non-null   int64         
 9   MntWines                  2216 non-null   int64         
 10  MntFruits                 2216 non-null   int64         
 11  MntMeatProducts           2216 non-null   int64         
 12  MntFishProducts   

## Remove unnecessary columns 

In [25]:
df.drop(columns = ['AcceptedCmp1'], inplace = True)
df.drop(columns = ['AcceptedCmp2'], inplace = True)
df.drop(columns = ['AcceptedCmp3'], inplace = True)
df.drop(columns = ['AcceptedCmp4'], inplace = True)
df.drop(columns = ['AcceptedCmp5'], inplace = True)

## Generate key
Generate key cho các bảng dimension

In [27]:
# Date_ID
df['Date_ID'] = df.apply(lambda x: str(x['Year_added']) + str(x['Month_added']) + str(x['Day_added']), axis=1)

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2216 entries, 1503 to 261
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   User_ID                   2216 non-null   int64         
 1   Year_Birth                2216 non-null   int64         
 2   Education                 2216 non-null   object        
 3   Marital_Status            2216 non-null   object        
 4   Income                    2216 non-null   float64       
 5   Kidhome                   2216 non-null   int64         
 6   Teenhome                  2216 non-null   int64         
 7   Date_added                2216 non-null   datetime64[ns]
 8   Recency                   2216 non-null   int64         
 9   MntWines                  2216 non-null   int64         
 10  MntFruits                 2216 non-null   int64         
 11  MntMeatProducts           2216 non-null   int64         
 12  MntFishProducts   

## Result
Dữ liệu sau khi được xử lý bao gồm 38 cột và 2216 dòng

In [29]:
df.head(5)

User_ID  Year_Birth   Education Marital_Status   Income  Kidhome  \
1503        0        1985  Graduation        Married  70951.0        0   
1           1        1961  Graduation         Single  57091.0        0   
1956        9        1975      Master         Single  46098.0        1   
1311       13        1947         PhD          Widow  25358.0        0   
1834       17        1971         PhD        Married  60491.0        0   

      Teenhome Date_added  Recency  MntWines  ...  NumStorePurchases  \
1503         0 2013-05-04       66       239  ...                  9   
1            0 2014-06-15        0       464  ...                  7   
1956         1 2012-08-18       86        57  ...                  2   
1311         1 2013-07-22       57        19  ...                  3   
1834         1 2013-09-06       81       637  ...                  7   

      NumWebVisitsMonth  Response  Complain       Country  \
1503                  1         0         0  South Africa   
1                     5         1         0        Canada   
1956                  8         0         0        Canada   
1311                  6         0         0  South Africa   
1834                  5         0         0         Spain   

      Total_Campaigns_Accepted  Year_added  Month_added  Day_added  Date_ID  
1503                         0        2013            5          4   201354  
1                            2        2014            6         15  2014615  
1956                         0        2012            8         18  2012818  
1311                         0        2013            7         22  2013722  
1834                         0        2013            9          6   201396  

[5 rows x 28 columns]

In [30]:
# lưu vào file csv
df.to_csv('dataset/marketing_data_final.csv')